File: src/main/java/com/dafttech/terra/game/world/subtiles/SubtileGrass.java
package com.dafttech.terra.game.world.subtiles;

import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.dafttech.terra.game.world.Vector2i;
import com.dafttech.terra.game.world.World;
import com.dafttech.terra.game.world.tiles.Tile;
import com.dafttech.terra.game.world.tiles.TileDirt;
import com.dafttech.terra.resources.Resources$;

import java.util.Random;

public class SubtileGrass extends Subtile {
    public SubtileGrass() {
        super();
    }

    @Override
    public TextureRegion getImage() {
        return Resources$.MODULE$.TILES().getImage("mask_grass");
    }

    int spreadDistance = 3;

    @Override
    public void onTick(World world, float delta) {
        if (world.getTile(tile.getPosition().addY(-1)).isOpaque()) {
            tile.removeSubtile(this);
        } else if (world.getTile(tile.getPosition().addY(-1)).getTemperature() > 50 && !(this instanceof SubtileDryGrass)) {
            tile.removeSubtile(this);
            tile.addSubtile(new SubtileDryGrass());
        } else {
            spread(world);
        }
    }

    public void spread(World world) {
        Vector2i spreadPosition = tile.getPosition().add(new Random().nextInt(spreadDistance * 2) - spreadDistance,
                new Random().nextInt(spreadDistance * 2) - spreadDistance);
        Tile spreadTile = world.getTile(spreadPosition);
        if (spreadTile != null && spreadTile instanceof TileDirt && world.getTile(spreadPosition.addY(-1)).isAir()) {
            if (!spreadTile.hasSubtile(SubtileGrass.class, true)) {
                spreadTile.addSubtile(new SubtileGrass());
            }
        }
    }
}

File: terra/src/main/scala/com/dafttech/terra/game/world/subtiles/SubtileGrass.scala
package com.dafttech.terra.game.world.subtiles

import java.util.Random

import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.dafttech.terra.engine.TilePosition
import com.dafttech.terra.game.world.tiles.TileDirt
import com.dafttech.terra.resources.Resources
import monix.eval.Task

class SubtileGrass extends Subtile {
  override val getImage: Task[TextureRegion] = Resources.TILES.getImageTask("mask_grass")

  private val spreadDistance: Int = 3

  override def onTick(delta: Float)(implicit tilePosition: TilePosition): Unit =
    if (tilePosition.world.getTile(tilePosition.pos - (0, 1)).tile.isOpaque) tilePosition.getTile.removeSubtile(this)
    else if (tilePosition.world.getTile(tilePosition.pos - (0, 1)).tile.getTemperature > 50 && !this.isInstanceOf[SubtileDryGrass]) {
      tilePosition.getTile.removeSubtile(this)
      tilePosition.getTile.addSubtile(new SubtileDryGrass)
    }
    else spread(tilePosition)

  def spread(tilePosition: TilePosition): Unit = {
    val spreadPosition = tilePosition.pos + (new Random().nextInt(spreadDistance * 2) - spreadDistance, new Random().nextInt(spreadDistance * 2) - spreadDistance)
    val spreadTile = tilePosition.world.getTile(spreadPosition)
    if (spreadTile != null && spreadTile.tile.isInstanceOf[TileDirt] && tilePosition.world.getTile(spreadPosition - (0, 1)).tile.isAir) if (!spreadTile.hasSubtile(classOf[SubtileGrass], true)) spreadTile.addSubtile(new SubtileGrass)
  }
}